ipq806x: add support for Askey RT4230W REV6

lmore377 opened the pull request:
This adds support for the Askey RT4230W REV6 (Branded by Spectrum/Charter as RAC2V1K)
At this time, there's no way to reinstall the stock firmware so don't install this on a router that's being rented.

Specifications:

    Qualcomm IPQ8065
    1 GB of RAM (DDR3)
    512 MB Flash (NAND)
    2x Wave 2 WiFi cards (QCA9984)
    5x 10/100/1000 Mbps Ethernet (Switch: QCA8337)
    1x LED (Controlled by a microcontroller that switches it between red and blue with different patterns)
    1x USB 3.0 Type-A
    12V DC Power Input
    UART header on PCB
        On ethernet port side; pinout from top to bottom is RX, TX, GND, 5V
        Port settings are 115200n8

More information: https://forum.openwrt.org/t/askey-rac2v1k-support/15830 https://deviwiki.com/wiki/Askey_RAC2V1K

To check what revision your router is, restore one of these config backups through the stock firmware to get ssh access then run "cat /proc/device-tree/model". https://forum.openwrt.org/t/askey-rac2v1k-support/15830/17
The revision number on the board doesn't seem to be very consistent, so that's why this is needed. You can also run printenv in the uboot console and if machid is set to 177d, that means your router is rev6.

Note: Don't install this if the router is being rented from an ISP. The defined partition layout is different from the OEM one and even if you redefined the layout to match, backing up and restoring the OEM firmware breaks /overlay so nothing will save and the router will likely enter a bootloop.

How to install:

<details>
<summary>Method 1: Install without opening the case using SSH and tftp</summary>
    
    You'll need:
    RAC2V1K-SSH.zip: https://github.com/lmore377/openwrt-rt4230w/blob/master/RAC2V1K-SSH.zip
    initramfs and sysupgrade images
    
    Connect to one of the router's LAN ports

    Download the RAC2V1K-SSH.zip file and restore the config file that corresponds to your router's firmware (If you're firmware is newer than what's in the zip file, just restore the 1.1.16 file)

    After a reboot, you should be able to ssh into the router with username: "4230w" and password: "linuxbox" or "admin". Run the following commannds
     fw_setenv ipaddr 10.42.0.10 #IP of router, can be anything
     fw_setenv serverip 10.42.0.1# #IP of tftp server that's set up in next steps
     fw_setenv bootdelay 8
     fw_setenv bootcmd "tftpboot initramfs.bin; bootm; bootipq"

    Don't reboot the router yet.

    Install and set up a tftp server on your computer

    Set a static ip on the ethernet interface of your computer (use this for serverip in the above commands)

    Rename the initramfs image to initramfs.bin, and host it with the tftp server

    Reboot the router. If you set up everything right, the router led should switch over to a slow blue glow which means openwrt is booted. If for some reason the file doesn't get loaded into ram properly, it should still boot to the OEM fw
    After openwrt boots, ssh into it and run these commands:
    fw_setenv bootcmd "setenv mtdids nand0=nand0 && setenv mtdparts mtdparts=nand0:0x1A000000@0x2400000(firmware) && ubi part firmware && ubi read 0x44000000 kernel 0x6e0000 && bootm"
    fw_setenv bootdelay 2

    After openwrt boots up, figure out a way to get the sysupgrade file onto it (scp, custom build with usb kernel module included, wget, etc.) then flash it with sysupgrade. After it finishes flashing, it should reboot, the light should go red for a bit then switch over to a blinking blue, then when the light starts "breathing" blue that means openwrt is booted.
</details>

<details>
<summary>Method 2: Install with serial access (Do this if something fails and you can't boot after using method 1) </summary>
    
    You'll need:
    initramfs and sysupgrade images
    Serial access: Instructions can be found here: https://openwrt.org/inbox/toh/askey/askey_rt4230w_rev6#opening_the_case

    Install and set up a tftp server

    Set a static ip on the ethernet interface of your computer

    Download the initramfs image, rename it to initramfs.bin, and host it with the tftp server

    Connect the wan port of the router to your computer

    Interrupt U-Boot and run these commands:
    setenv serverip 10.42.0.1 (You can use whatever ip you set for the computer)
    setenv ipaddr 10.42.0.10 (Can be any ip as long as it's in the same subnet)
    setenv bootcmd "setenv mtdids nand0=nand0 && set mtdparts mtdparts=nand0:0x1A000000@0x2400000(firmware) && ubi part firmware && ubi read 0x44000000 kernel 0x6e0000 && bootm"

    saveenv
    tftpboot initramfs.bin
    bootm
    
    After openwrt boots up, figure out a way to get the sysupgrade file onto it (scp, custom build with usb kernel module included, wget, etc.) then flash it with sysupgrade. After it finishes flashing, it should reboot, the light should go red for a bit then switch over to a blinking blue, then when the light starts "breathing" blue that means openwrt is booted.
</details>

adrianschmutzler commented:
> This adds support for the Askey RT4230W REV6 (Branded by Spectrum/Charter as RAC2V1K)
> This router has an almost identical twin (Branded by Spectrum/Charter as SAC2V1K) that has 256MiB of flash.

Does that mean that both RAC2V1K and SAC2V1K are an Askey RT4230W REV6?

If that's the case, i.e. using Askey RT4230W REV6 would not allow to discriminate between both versions, I'd strongly recommend to use the brand names instead.

adrianschmutzler commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -316,4 +316,21 @@ define Device/zyxel_nbg6817
 endef
 TARGET_DEVICES += zyxel_nbg6817
 
+define Device/askey_rt4230w
```
What happened to the revision here?

lmore377 replied:
I want to keep it clean so I ended up putting the revision in the device name. Also I tried adding _rev6 to that line, but then make started looking for a dts file named qcom-ipq8065-rev6.dts when trying to build.

lxl750629 replied:
Excuse me, where can I download initramfs image?

lmore377 replied:
Image and instructions for installing to nand here: https://github.com/lmore377/openwrt-rt4230w

lmore377 replied:
> Excuse me, where can I download initramfs image?
@lxl750629
Image and instructions for installing to nand here: https://github.com/lmore377/openwrt-rt4230w
Discussion is here: https://forum.openwrt.org/t/askey-rac2v1k-support/15830

adrianschmutzler commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -316,4 +316,21 @@ define Device/zyxel_nbg6817
+	$(call Device/LegacyImage)
+	$(call Device/UbiFit)
+        DEVICE_VENDOR := Askey
+        DEVICE_MODEL := RAC2V1K/RT4230W
```
Which is it? Decide for one name (either consistently use board name or branded name).
If board name is not unique for the setup, it will be branded name.

lmore377 replied:
Decided on RT4230W-REV6 as the device name.

adrianschmutzler commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -316,4 +316,21 @@ define Device/zyxel_nbg6817
+        SOC := qcom-ipq8065
+        BLOCKSIZE := 128k
+        PAGESIZE := 2048
+        BOARD_NAME := rt4230w
```
That's probably not required.

lmore377 replied:
I left it because deleting it broke installation

adrianschmutzler commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -316,4 +316,21 @@ define Device/zyxel_nbg6817
+TARGET_DEVICES += askey_rt4230w
+
+
+
```
One empty line should be enough.

adrianschmutzler commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -316,4 +316,21 @@ define Device/zyxel_nbg6817
+        PAGESIZE := 2048
+        BOARD_NAME := rt4230w
+        IMAGE/factory.bin := append-kernel | append-ubi
+        IMAGE/nand-sysupgrade.bin := sysupgrade-tar | append-metadata
```
please use indent as for the other devices.

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w.dts:
```diff
@@ -0,0 +1,373 @@
+&tcsr {
+	compatible = "qcom,tcsr";
+	reg = <0x1a400000 0x100>;
+	status = "ok";
```
please use "okay".

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w.dts:
```diff
@@ -0,0 +1,373 @@
+			gpios = <&qcom_pinmux 23 GPIO_ACTIVE_HIGH>;
+		};
+
+		ledctrl3: ledctrl3 {
```
Are the DT labels needed?

lmore377 replied:
Yes because I'm using 2 and 3 for boot status

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w.dts:
```diff
@@ -0,0 +1,373 @@
+		};
+
+		ledctrl3: ledctrl3 {
+			label = "rt4230w:ledctrl:3";
```
Typically, LED labels are set up to a different logic. Maybe have look at some other DTS files here ...

lmore377 replied:
I set up the leds like this because based on how you change the 3 gpio pins, it does different things. For example:
on,off,off = red glow
on,off,on = red/blue glow
off,on,on = solid blue
etc.
A script needs to be made to control the led properly, unless there's a way to define multiple gpio pins under one led.

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w.dts:
```diff
@@ -0,0 +1,373 @@
+			label = "rt4230w:ledctrl:3";
+			gpios = <&qcom_pinmux 24 GPIO_ACTIVE_HIGH>;
+		};
+
```
Remove useless empty lines. Empty lines are typically just needed between nodes of the same level.

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w.dts:
```diff
@@ -0,0 +1,373 @@
+		compatible = "gpio-leds";
+		pinctrl-0 = <&led_pins>;
+		pinctrl-names = "default";
+		ledctrl1: ledctrl1 {
```
Add empty line before the first led node here.

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w.dts:
```diff
@@ -0,0 +1,373 @@
+			phy-mode = "sgmii";
+			qcom,id = <1>;
+
+			mtd-mac-address = <&ART 0>;
```
0x0 is preferred over 0 here.

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w.dts:
```diff
@@ -0,0 +1,373 @@
+			pinctrl-0 = <&rgmii2_pins>;
+			pinctrl-names = "default";
+
+			mtd-mac-address = <&ART 6>;
```
0x6 is preferred over 6 here.

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w.dts:
```diff
@@ -0,0 +1,373 @@
+						reg = <0x0000000 0x0040000>;
+						read-only;
+					};
+					MIBIB@0 {
```
The value after "@" should be the hex of the partition start.

Apart from that, generic names are preferred, e.g. partition@0, partition@40000, ....

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w.dts:
```diff
@@ -0,0 +1,373 @@
+						reg = <0x1180000 0x0080000>;
+						read-only;
+					};
+					ART: ART@1200000 {
```
Consider using lower-case "art" for the DT label and the label property here.

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w.dts:
```diff
@@ -0,0 +1,373 @@
+			 */
+		};
+
+
```
No double empty lines.

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w.dts:
```diff
@@ -0,0 +1,373 @@
+	
+#include <dt-bindings/input/input.h>
+
+
```
No double empty lines.

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w.dts:
```diff
@@ -0,0 +1,373 @@
+#include "qcom-ipq8065.dtsi"
```
Please add an SPDX license identifier ...

adrianschmutzler commented on target/linux/ipq806x/base-files/lib/upgrade/platform.sh:
```diff
@@ -40,6 +40,10 @@ platform_do_upgrade() {
 		MTD_CONFIG_ARGS="-s 0x200000"
 		default_do_upgrade "$1"
 		;;
+	askey,rt4230w)
```
please take care of alphabetic sorting.

adrianschmutzler commented on target/linux/ipq806x/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -73,6 +77,10 @@ case "$FIRMWARE" in
 		caldata_extract "0:ART" 0x5000 0x2f20
 		ath10k_patch_mac $(mtd_get_mac_ascii 0:APPSBLENV ethaddr)
 		;;
+	askey,rt4230w)
```
Sorting.

adrianschmutzler commented on target/linux/ipq806x/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -73,6 +77,10 @@ case "$FIRMWARE" in
 		;;
+	askey,rt4230w)
+		caldata_extract "ART" 0x5000 0x7f1f
+		ath10k_patch_mac $(macaddr_add $(mtd_get_mac_binary ART 0x6) +2)
```
No address found in ART 0x5006?

lmore377 replied:
The MAC addresses match up with stock fw. Confirmed with ifconfig and WiFi Analyzer on android.

adrianschmutzler commented on target/linux/ipq806x/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -39,6 +39,10 @@ case "$FIRMWARE" in
 		;;
+	askey,rt4230w)
+		caldata_extract "ART" 0x1000 0x3f1f
+		ath10k_patch_mac $(macaddr_add $(mtd_get_mac_binary ART 0x6) +1)
```
No address found in ART 0x1002?

lmore377 replied:
The MAC addresses match up with stock fw. Confirmed with ifconfig and WiFi Analyzer on android.

adrianschmutzler commented on target/linux/ipq806x/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -39,6 +39,10 @@ case "$FIRMWARE" in
 		caldata_extract "0:ART" 0x1000 0x2f20
 		ath10k_patch_mac $(macaddr_add $(mtd_get_mac_ascii 0:APPSBLENV ethaddr) +1)
 		;;
+	askey,rt4230w)
```
Sorting.

adrianschmutzler commented on target/linux/ipq806x/base-files/etc/board.d/02_network:
```diff
@@ -55,6 +55,10 @@ zyxel,nbg6817)
 	ucidef_set_interface_macaddr "lan" "$(macaddr_add $hw_mac_addr 2)"
 	ucidef_set_interface_macaddr "wan" "$(macaddr_add $hw_mac_addr 3)"
 	;;
+askey,rt4230w)
```
Sorting.

adrianschmutzler commented on target/linux/ipq806x/base-files/etc/board.d/02_network:
```diff
@@ -55,6 +55,10 @@ zyxel,nbg6817)
 	;;
+askey,rt4230w)
+	ucidef_add_switch "switch0" \
+		"1:wan" "2:lan" "3:lan" "4:lan" "6@eth1" "5:lan" "0@eth0"
```
If you don't care about the port order anyway, please put the ethX entries in a way that they separate lan from wan ports ...

lmore377 commented:
> > This adds support for the Askey RT4230W REV6 (Branded by Spectrum/Charter as RAC2V1K)
> > This router has an almost identical twin (Branded by Spectrum/Charter as SAC2V1K) that has 256MiB of flash.
> 
> Does that mean that both RAC2V1K and SAC2V1K are an Askey RT4230W REV6?
> 
> If that's the case, i.e. using Askey RT4230W REV6 would not allow to discriminate between both versions, I'd strongly recommend to use the brand names instead.

According to the DTS files I extracted from the stock fw, there's a rev 1,3,5,6, and 10. 6 and 10 are identical except for the smaller flash on 10 but 1,3, and 5 have different gpios and/or different ethernet configurations. I'm pretty sure those other revisions are under the same brand name so I'd rather go by the rev number reported by the oem firmware.

adrianschmutzler commented:
So effectively, all revisions are different, so they need to be included in the name.

However, that doesn't sound like the SAC will be the same revision as the RAC?

lmore377 commented:
> So effectively, all revisions are different, so they need to be included in the name.
> 
> However, that doesn't sound like the SAC will be the same revision as the RAC?
The RAC I have is rev6 and I also have a SAC and that's rev10 but I don't want to mess with it because it's rented

lmore377 commented:
Is there anything else that needs fixing?

adrianschmutzler commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -81,6 +81,21 @@ define Device/ZyXELImage
 	IMAGE/sysupgrade.bin/squashfs := append-rootfs | pad-to $$$$(BLOCKSIZE) | sysupgrade-tar rootfs=$$$$@ | append-metadata
 endef
 
+define Device/askey_rt4230w
```
This should be
askey_rt4230w-rev6
then.

The compatible in DTS and the base-files case entries should be updated to askey,rt4230w-rev6 then.

adrianschmutzler commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -81,6 +81,21 @@ define Device/ZyXELImage
+	$(call Device/LegacyImage)
+	$(call Device/UbiFit)
+	DEVICE_VENDOR := Askey
+	DEVICE_MODEL := RT4230W-REV6
```
DEVICE_MODEL := RT4230W
DEVICE_VARIANT := Rev6

adrianschmutzler commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -81,6 +81,21 @@ define Device/ZyXELImage
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	BOARD_NAME := rt4230w
+	DEVICE_PACKAGES := ath10k-firmware-qca9984-ct-full-htt
```
Based on the other devices in this target, it looks to me like you should choose ath10k-firmware-qca9984-ct here (without -full-htt). Or do you need the latter for some specific reason?

lmore377 replied:
Just needed -ct. That was left over from me trying to figure out why I couldn't change the 5Ghz channel but it turns out it was just taking a bit to bring up the interface.

adrianschmutzler commented on target/linux/ipq806x/base-files/etc/board.d/02_network:
```diff
@@ -22,6 +22,10 @@ tplink,vr2600v)
 	ucidef_add_switch "switch0" \
 		"1:lan" "2:lan" "3:lan" "4:lan" "6@eth1" "5:wan" "0@eth0"
 	;;
+askey,rt4230w)
```
askey comes before buffalo ...

adrianschmutzler replied:
Despite, this can be merged with nec,wg2600hp; so create one case entry for both of them and move it up.

adrianschmutzler commented on target/linux/ipq806x/base-files/etc/board.d/02_network:
```diff
@@ -55,6 +59,10 @@ zyxel,nbg6817)
 	;;
+askey,rt4230w)
+	ucidef_add_switch "switch0" \
+		"1:wan" "2:lan" "3:lan" "4:lan" "6@eth1" "5:lan" "0@eth0"
```
Currently, this is done twice with different assignment ...

adrianschmutzler commented on target/linux/ipq806x/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -9,6 +9,10 @@ board=$(board_name)
 	case $board in
+	askey,rt4230w)
+		caldata_extract "art" 0x1000 0x3f1f
+		ath10k_patch_mac $(macaddr_add $(mtd_get_mac_binary ART 0x6) +1)
```
So, the addresses are correct, thanks. I'd be interested whether there are addresses in 0x1002 and 0x5006 as well.
Can you give a table with
0x0
0x6
0x1002
0x5006
?

The last byte of the addresses would be enough.

Further reading: https://openwrt.org/docs/guide-developer/mac.address#find_out_about_flash_locations

lmore377 replied:
oh god the last byte of the mac addresses for wireless don't match up. Umm give me a few minutes to fix that.

lmore377 replied:
Turns out it was originally getting the ethernet mac addresses and adding 1 or 2 to the last byte. The last byte is the only different one between ethernet and wifi mac addresses so I didn't catch that.

adrianschmutzler commented on target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8065-rt4230w.dts:
```diff
@@ -0,0 +1,372 @@
+#include <dt-bindings/input/input.h>
+
+/ {
+	model = "Askey RT4230W";
```
Rev6?

adrianschmutzler commented on target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8065-rt4230w.dts:
```diff
@@ -0,0 +1,372 @@
+		serial0 = &gsbi4_serial;
+		mdio-gpio0 = &mdio0;
+		led-boot = &ledctrl3;
+		led-running = &ledctrl2;
```
What about failsafe and upgrade indication?

lmore377 replied:
Is having boot and upgrade the same fine? Also is a red glow for failsafe fine?

adrianschmutzler commented on target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8065-rt4230w.dts:
```diff
@@ -0,0 +1,372 @@
+					#address-cells = <1>;
+					#size-cells = <1>;
+
+					partition0@0 {
```
AFAIK you can drop the number _before_ the "@", just "partition@0"

adrianschmutzler commented:
When you are done, please squash everything back together into a single proper commit (git rebase -i ...) and force-push it to your GitHub branch again (git push -f ...).

lmore377 commented:
> When you are done, please squash everything back together into a single proper commit (git rebase -i ...) and force-push it to your GitHub branch again (git push -f ...).

I've been editing things through the github web interface

adrianschmutzler commented on target/linux/ipq806x/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -11,7 +11,7 @@ case "$FIRMWARE" in
 	askey,rt4230w-rev6)
 		caldata_extract "art" 0x1000 0x3f1f
-		ath10k_patch_mac $(macaddr_add $(mtd_get_mac_binary ART 0x6) +1)
+		ath10k_patch_mac $(macaddr_add $(mtd_get_mac_binary ART 0x1006))
```
If you don't increment anymore, you can drop the macaddr_add as well:
ath10k_patch_mac $(mtd_get_mac_binary ART 0x1006)

Same for the other one ...

adrianschmutzler replied:
Err, btw, you now have the default position inside calibration data, so you should get correct MAC addresses even if you drop the ath10k_patch_mac lines entirely, as caldata_extract will read them from that location anyway. Please test!

adrianschmutzler commented on target/linux/ipq806x/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -49,7 +49,7 @@ case "$FIRMWARE" in
 	askey,rt4230w-rev6)
 		caldata_extract "art" 0x5000 0x7f1f
-		ath10k_patch_mac $(macaddr_add $(mtd_get_mac_binary ART 0x6) +2)
+		ath10k_patch_mac $(macaddr_add $(mtd_get_mac_binary ART 0x5006))
```
See above: If you read from 0x5006 here, you should be able to remove the second line entirely.

lmore377 replied:
Yeah it worked. I was just going off what all the other devices did

lmore377 commented:
Also, I've been testing everything with the 4.19 kernel because using 5.4 just kernel panics. Do I need to get that working before everything's finalized?

adrianschmutzler commented:
> Also, I've been testing everything with the 4.19 kernel because using 5.4 just kernel panics. Do I need to get that working before everything's finalized?

The target is on 5.4, so if the device does not work with 5.4 we cannot merge it.

lmore377 commented:
> > Also, I've been testing everything with the 4.19 kernel because using 5.4 just kernel panics. Do I need to get that working before everything's finalized?
> 
> The target is on 5.4, so if the device does not work with 5.4 we cannot merge it.

Alright. Good thing is I got it booting and stable, bad thing is pcie is broken completely. I just get this for both controllers: `PCIe controller is not set to bridge type (hdr_type: 0xff)!`

Do I keep the request open until I get it working or do I need to do something else?

lmore377 commented:
I got it working on 5.4, but it uses a workaround in the stock uboot to boot. Is this fine as long as I provide instructions on how to do it?

lmore377 commented:
@adrianschmutzler Any word on if this can be approved or not?

adrianschmutzler commented:
I'm only reviewing parts of this, I won't merge it.

However, in order to have somebody else looking at it, please squash all your commits into a single proper commit again.

mans0n commented:
> I just get this for both controllers: `PCIe controller is not set to bridge type (hdr_type: 0xff)!`

I've just encountered the same error and I'm searching how to resolve it. Could you tell me how you got it working?

I also would like to review your patch, I'd be happy if you squash the commits and tidy up indents first...

lmore377 commented:
@mans0n can you tell me how you fixed your problem in the dts? Also uboot is related to this because the original firmware on this router was launched with the bootipq command and I don't know what it does but when I don't run it I get kernel warnings about regulators and pcie doesn't work but when I do run it, it fails because the original firmware is missing but if I boot openwrt after that with tftp or just loading directly from the nand, it just runs perfectly fine.

mans0n commented:
Interesting, maybe bootipq does initialization stuff?

In my case the culprit was pcie reset. Try workarounds used in d7800 or wpq864. I used:
```
&pcie0 {
	status = "okay";

	reset-gpio = <&qcom_pinmux 3 GPIO_ACTIVE_HIGH>;
	/delete-property/ perst-gpios;
};

&pcie1 {
	status = "okay";

	reset-gpio = <&qcom_pinmux 48 GPIO_ACTIVE_HIGH>;
	/delete-property/ perst-gpios;
};
```

lmore377 commented:
@mans0n This worked! Thank you so much! I had been looking for a fix but I just couldn't figure out was wrong. No more having to use a bootcmd that relies on bootipq failing lol

lmore377 commented:
I'm going to recompile and test everything one last time before squashing everything together

lmore377 commented:
@mans0n I'm pretty sure I squashed everything together properly and fixed the indents so everything should be good to go. I'm just building *one last time* to make sure it's all good

mans0n commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,379 @@
+
+&gmac1 {
+	status = "okay";
+			phy-mode = "sgmii";
```
By "indent" I mean these lines (not only this line). Compare your patch to other dts files and you'll know what to improve.

lmore377 replied:
I shouldn't do late night editing but it's like the only time I can

mans0n commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,379 @@
+&mdio0 {
+	status = "okay";
+
+	compatible = "virtual,mdio-gpio";
```
And you don't need to use mdio-gpio anymore (in 5.4).

adschm replied:
Why did you mark this as resolved?

mans0n replied:
Is this resolved? https://github.com/openwrt/openwrt/pull/3115#discussion_r446612461
Please have a look to clean this up. https://github.com/openwrt/openwrt/blob/master/target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8064-eax500.dtsi#L154
I believe you don't need `ethernet-phy@4` node either.
If you succeed in removing `compatible = "virtual,mdio-gpio";`, then you can also drop `mdio-gpio0 = &mdio0;` line in the `aliases` node.

lmore377 replied:
I was able to remove everything mentioned and it still works fine so I'll mark this as resolved.

mans0n commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -81,6 +81,22 @@ define Device/ZyXELImage
+	SOC := qcom-ipq8065
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	BOARD_NAME := rt4230w-rev6
```
If you get a warning during sysupgrade when you remove this line, it might be because your (old) installed system is already using it. Please try forcing sysupgrade once and check out it sill requires BOARD_NAME.

lmore377 replied:
Do you mean just remove the BOARD_NAME line entirely?

mans0n replied:
Right, it should not be needed on newly added devices.

mans0n commented:
Please wrap up long lines in the commit message. You don't need the lines after Signed-off-by

lmore377 commented:
I think I did everything properly this time.

lxl750629 commented:
Copy that. Thank you very much!




------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"lmore377"<notifications@github.com&gt;;
发送时间:&nbsp;2020年6月29日(星期一) 上午10:50
收件人:&nbsp;"openwrt/openwrt"<openwrt@noreply.github.com&gt;;
抄送:&nbsp;"68405038"<68405038@qq.com&gt;;"Comment"<comment@noreply.github.com&gt;;
主题:&nbsp;Re: [openwrt/openwrt] ipq806x: add support for Askey RT4230W REV6 (#3115)





 
@lmore377 commented on this pull request.
 
 
In target/linux/ipq806x/image/Makefile:
 &gt; @@ -316,4 +316,21 @@ define Device/zyxel_nbg6817  endef  TARGET_DEVICES += zyxel_nbg6817   +define Device/askey_rt4230w  
Image and instructions for installing to nand here: https://github.com/lmore377/openwrt-rt4230w
 
—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or unsubscribe.

lmore377 commented:
@mans0n I'm pretty sure I got everything in order now

lxl750629 commented:
Is there a way to replace the UBOOT of RAC2V1K with the UBOOT? of Netgear R7800? In this way, it will be much more convenient and secure to update the firmware in the future.

lmore377 commented:
Why would it be more secure? I've done plenty of sysupgrades while working on this and never ran into issues.

lxl750629 commented:
Thank you. I have installed openwrt according to your tutorial.&nbsp;If I want to install my own compiled RAC2V1K firmware which is different from the REV6 partition format provided by you (UBOOT), including netgear R7800), how do I install it?&nbsp;Can I install it according to the existing steps?&nbsp;Thank you!

lmore377 commented:
It depends on what exactly you changed but in general, using sysupgrade in an initramfs image should work fine

mans0n commented:
I found out that the pcie driver does not use reset-gpio property. I could safely remove it and keep `/delete-property/ perst-gpios;` only.

mans0n commented on target/linux/ipq806x/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -9,6 +9,9 @@ board=$(board_name)
 "ath10k/pre-cal-pci-0000:01:00.0.bin")
 	case $board in
+	askey,rt4230w-rev6)
+		caldata_extract "art" 0x1000 0x3f1f
```
Your extract size is different from the other devices. Are you sure you need that much?

lmore377 replied:
I thought I needed to define the start and end of each one

adrianschmutzler replied:
It's start and size.

mans0n commented on target/linux/ipq806x/base-files/lib/upgrade/platform.sh:
```diff
@@ -10,6 +10,10 @@ platform_check_image() {
 platform_do_upgrade() {
 	case "$(board_name)" in
+	askey,rt4230w-rev6)
+		PART_NAME="ubi"
```
PART_NAME is for default_do_upgrade. nand_do_upgrade uses CI_UBIPART, and the default value is ubi. So probably you can remove this line.

adschm replied:
@lmore377 Why did you mark this as resolved?

lmore377 replied:
I removed it and sysupgrade still works.

mans0n commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -81,6 +81,21 @@ define Device/ZyXELImage
+	PAGESIZE := 2048
+	DEVICE_PACKAGES := ath10k-firmware-qca9984-ct
+	KERNEL_IN_UBI := 1
+	IMAGES := nand-sysupgrade.bin
```
Where is the factory image mentioned in the commit message?

lmore377 replied:
I switched to using sysupgrade and forgot to change that

mans0n replied:
So you don't use factory image anymore during installation. But still nand-factory ubi image might be useful for someone else (unless the image is totally absurd for this device). I guess you can just use `$(call Device/UbiFit)` here.

lmore377 commented:
> I found out that the pcie driver does not use reset-gpio property. I could safely remove it and keep `/delete-property/ perst-gpios;` only.

Should I leave or keep it?

mans0n commented:
>> I found out that the pcie driver does not use reset-gpio property. I could safely remove it and keep /delete-property/ perst-gpios; only.
>
> Should I remove or keep it?

IMO it's okay to remove reset-gpio, if that works for you. I'm not sure why other devices used it...

Ansuel commented:
Hey can you test this??

```&pcie0 {
	status = "okay";

	perst-gpio = <&qcom_pinmux 3 GPIO_ACTIVE_HIGH>;
	/delete-property/ perst-gpios;
};

&pcie1 {
	status = "okay";

	/delete-property/ perst-gpios;
};```

mans0n commented:
> Hey can you test this??

I've actually tried it (using GPIO_ACTIVE_HIGH for perst-gpio) first and it did reset the bus, but the wlan chip did not wake up. So I had to disable the reset pin at all.

Ansuel commented:
The definition is different and also only the first rst gpio pin is different. 
There are 3 reference board with 3 different configuration about this damm gpio reset

lxl750629 commented:
Okay!&nbsp;Could you tell me how to take the test?




------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"mans0n"<notifications@github.com&gt;;
发送时间:&nbsp;2020年7月1日(星期三) 凌晨0:32
收件人:&nbsp;"openwrt/openwrt"<openwrt@noreply.github.com&gt;;
抄送:&nbsp;"68405038"<68405038@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [openwrt/openwrt] ipq806x: add support for Askey RT4230W REV6 (#3115)





  
Hey can you test this??
  
I've actually tried it first (using GPIO_ACTIVE_HIGH for perst-gpio) and it did reset the bus, but the wlan chip did not wake up. So I had to disable the reset pin at all.
 
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.

lmore377 commented:
I'm not sure if this can help with the gpio reset thing at all but here are some dts files I extracted from my router https://github.com/lmore377/openwrt-rt4230w/tree/master/dts-files

lxl750629 commented:
After openwrt has been installed, how can I restore it to the original factory system?

lmore377 commented:
There's no way to currently. That was one of the first things I was trying to do and I had the MTD layout defined like it was in the oem firmware and was able to dump an image but after I reflashed it, it broke the ubi partition so it would boot but uboot would throw lots of errors and all changes would be lost after a reboot.

lmore377 commented:
Can this pull request to be approved yet or is there still stuff I need to fix?

lmore377 commented:
@mans0n

mans0n commented:
Sorry, your patch and the commit message still needs more formatting (such as `qca,ar8327-initvals` or 80-column per line), and I'm having a difficulty reviewing it...

adschm commented on target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,373 @@
+	};
+
+	aliases {
+		serial0 = &gsbi4_serial;
```
serial0 can be dropped, it's in the DTSI now.

adschm replied:
Sorry, this comment applies to 5.4 only.

adschm commented on target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,373 @@
+
+	chosen {
+		bootargs = "rootfstype=squashfs noinitrd";
+		stdout-path = "serial0:115200n8";
```
5.4 only: stdout-path can be dropped, it's in the DTSI now.

adschm commented on target/linux/ipq806x/base-files/etc/board.d/02_network:
```diff
@@ -41,9 +46,12 @@ linksys,ea8500)
-		"2:lan" "3:lan" "4:lan" "5:lan" "6@eth1" "1:wan" "0@eth0"
+		"1:lan" "2:lan" "3:lan" "4:lan" "6@eth1" "5:wan" "0@eth0"
+	ucidef_set_interface_macaddr "lan" "$hw_mac_addr"
+	ucidef_set_interface_macaddr "wan" "$hw_mac_addr"
```
This looks broken.

lmore377 replied:
Fixed it

adschm commented on target/linux/ipq806x/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -44,6 +47,9 @@ case "$FIRMWARE" in
 "ath10k/pre-cal-pci-0001:01:00.0.bin")
 	case $board in
+	askey,rt4230w-rev6)
+		caldata_extract "art" 0x5000 0x7f1f
```
Note that the third parameter is the size. Looks to me like you should just use 0x2f20 like all the others.

lmore377 replied:
I thought you had to define the start and end. Fixed

adschm commented on target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,373 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "qcom-ipq8065.dtsi"
+#include <dt-bindings/input/input.h>
```
please just drop the 4.19 files entirely to make this easier to review.

lmore377 replied:
I only have the 5.4 files now

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,379 @@
+			reg = <0x41200000 0x300000>;
+			no-map;
+		};
+	};
```
This whole block can be dropped due to https://github.com/openwrt/openwrt/commit/57e029ba573bb6b4e19c90d55960bed0357c321b

lmore377 replied:
Dropped

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,379 @@
+	};
+
+	aliases {
+		serial0 = &gsbi4_serial;
```
The serial0 line can be dropped, it's in the dtsi.

lmore377 replied:
Droopped

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,379 @@
+
+	chosen {
+		bootargs = "rootfstype=squashfs noinitrd";
+		stdout-path = "serial0:115200n8";
```
The stdout-path can be dropped, it's in the dtsi.

lmore377 replied:
Droopped

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,379 @@
+		pinctrl-names = "default";
+
+		ledctrl1: ledctrl1 {
+			label = "rt4230w:ledctrl:1";
```
Please use a proper naming scheme for the LEDs:
https://openwrt.org/docs/guide-developer/device-support-policies#leds

Based on that, please update node names and DT labels.

lmore377 replied:
I honestly still don't know how to deal with these. Most routers have seperate leds but this one just has a red/blue led connected to a microcontroller. [You can see what I mean here](https://github.com/openwrt/openwrt/pull/3115/#discussion_r441098331)

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,379 @@
+			drive-strength = <12>;
+			bias-disable;
+		};
+	};
```
i2c4_pins can be removed here, it's in the dtsi now.

lmore377 replied:
Droopped

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,379 @@
+			drive-strength = <10>;
+			bias-bus-hold;
+		};
+	};
```
nand_pins can be removed here, it's in the dtsi now.

lmore377 replied:
Dropped

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,379 @@
+			drive-strength = <8>;
+			bias-disable;
+		};
+	};
```
mdio_pins can be removed here, it's in the dtsi now.

lmore377 replied:
Dropped

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,379 @@
+			function = "rgmii2";
+			drive-strength = <8>;
+			bias-disable;
+		};
```
Is gpio32 missing here intentionally? If not, you can drop the mux {} node, and just keep the tx here.

lmore377 replied:
Yeah it is. This router is weird and for some reason it has 2 transistors that are hooked up to that pin. When it's pulled low it disables the wifi cards until it's rebooted.

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,379 @@
+	* RPM usage. Turning this on in kernel manifests as
+	* i2c failure for the RPM.
+	*/
+};
```
This whole node can be dropped, it has been moved to the dtsi.

lmore377 replied:
Dropped

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,379 @@
+		phy0: ethernet-phy@0 {
+		device_type = "ethernet-phy";
+		reg = <0x0>;
+		qca,ar8327-initvals = <0x04 0x7600000 0x08 0x1000000 0x0c 0x80 0xe4 0xaa545 0xe0 0xc74164de 0x7c 0x4e 0x94 0x4e 0x50 0xcf02cf02 0x54 0xc832c832>;
```
Please wrap this line. Despite, this node has broken indent.

lmore377 replied:
Wrapped and fixed indent

adschm commented:
Changes to 0069-arm-boot-add-dts-files.patch are missing.

Please don't just mark stuff as resolved, but answer the questions.

lmore377 commented:
I barely got the chance to take a look at this again (school started and I was using the router as my main). I'm going to try to clean up my branch and get it in sync with this repo before making anymore changes.

lmore377 commented:
Alright to be honest I don't know what I did and things broke. I'm just going to start from scratch and try to fix everything up. It'll probably be 2-3 weeks before this is in a good state. Feel free to do whatever needs to be done to this pr in the meantime (if anything).

lmore377 commented:
Nevermind I got it all good faster than I expected.

lmore377 commented:
> Changes to 0069-arm-boot-add-dts-files.patch are missing.
> 

Can you help me out with editing that? I got it working by changing `+843,19` to `+843,20` and adding `qcom-ipq8065-rt4230w-rev6.dtb` right above `qcom-ipq8065-nbg6817.dtb` but I'm almost certain that's not the right way to do it.

lmore377 commented:
@adschm Can you take another look over everything? I think I have pretty much everything except  0069-arm-boot-add-dts-files.patch fixed at this point.

lmore377 commented:
@mans0n can you take another look at this? Also, what's your opinion on what I should do with the LED names? On this router it's one bi-color LED that flashes different patterns based on which gpio pins are toggled on and off (you can see what I mean on[ this issue](https://github.com/openwrt/openwrt/pull/3115#discussion_r441098331))

mans0n commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,317 @@
+				reg = <0x1180000 0x0080000>;
+			};
+			art: partition@1200000 {
+				label = "art";
```
Is there a reason you chose to use lowercase "art" here? It seems all other system partitions use uppercase labels.

lmore377 replied:
There's no particular reason. Should I leave it as-is or make it uppercase?

lmore377 replied:
Just ended up making it uppercase.

mans0n commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,317 @@
+#include <dt-bindings/input/input.h>
+
+/ {
+	model = "Askey RT4230W";
```
Why doesn't it end with "REV6"?

lmore377 replied:
Just a typo I hadn't noticed. I'll mark this along with other things as resolved once I test everything and push a commit

mans0n commented:
> on,off,off = red glow
> on,off,on = red/blue glow
> off,on,on = solid blue

Does that mean line 1 is red, line 2 is blink/solid control, and line 3 is blue?

mans0n commented:
> > Changes to 0069-arm-boot-add-dts-files.patch are missing.
>
> Can you help me out with editing that? I got it working by changing +843,19 to +843,20 and adding qcom-ipq8065-rt4230w-rev6.dtb right above qcom-ipq8065-nbg6817.dtb but I'm almost certain that's not the right way to do it.

That is enough. https://github.com/openwrt/openwrt/commit/20c7abd4b7b9b4203b6858136b2c811b6dbd8dae#diff-d174985bfe7810654ec2d3c106d73ab3R21
BTW does this patch even build without the 0069.patch changes?

lmore377 commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,317 @@
+			gpios = <&qcom_pinmux 54 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+			debounce-interval = <60>;
+			wakeup-source;
```
I know what debouncing is but I'll just remove those lines anyways

lmore377 commented on target/linux/ipq806x/base-files/etc/board.d/02_network:
```diff
@@ -41,6 +41,7 @@ linksys,ea8500)
 	ucidef_set_interface_macaddr "lan" "$hw_mac_addr"
 	ucidef_set_interface_macaddr "wan" "$hw_mac_addr"
 	;;
+askey,rt4230w-rev6 |\
```
Sorted

adschm replied:
And now move the block based on proper sorting.

lmore377 replied:
Moved it

lmore377 commented:
> Does that mean line 1 is red, line 2 is blink/solid control, and line 3 is blue?

Nope. Here's a list of what the led does with different combos of the gpio pins. By breathing I mean that effect where a light fades in and out
off,off,off - off
on,off,off - breathing red
on,on,off - solid red
on,on,on - pink (blue and red)
off,on,on - solid blue
off,off,on - blinking blue
on,off,on - breathing between red and blue
off,on,off - breathing blue

>  BTW does this patch even build without the 0069.patch changes?

Yeah it builds fine and I don't notice anything wrong

lmore377 commented:
Do I need to fix the conflicts or will they be fixed later? In the uboot file the issue is that support for another device was added at some point and in platform.sh the issue is the same as the uboot file but I also moved a part of the file since the command for my device was the same as some others

adschm commented on target/linux/ipq806x/base-files/lib/upgrade/platform.sh:
```diff
@@ -5,44 +5,45 @@ RAMFS_COPY_BIN='fw_printenv fw_setenv'
+        *)
+                default_do_upgrade "$1"
+                ;;
+        esac
```
Please don't change indent here.

lmore377 replied:
I don't even know how that happened but it's fixed now

adschm commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -82,6 +82,21 @@ define Device/ZyXELImage
+        DEVICE_PACKAGES := ath10k-firmware-qca9984-ct
+        KERNEL_IN_UBI := 1
+        IMAGES := nand-sysupgrade.bin
+        IMAGE/nand-sysupgrade.bin := sysupgrade-tar | append-metadata
```
Please use tabs for indent here like the other devices do.

lmore377 replied:
Done

adschm commented on package/boot/uboot-envtools/files/ipq806x:
```diff
@@ -31,6 +31,9 @@ ubootenv_mtdinfo () {
 case "$board" in
+askey,rt4230w-rev6)
+  ubootenv_add_uci_config "/dev/mtd9" "0x0" "0x40000" "0x20000"
+  ;;
```
Broken indent.

adschm commented on target/linux/ipq806x/base-files/lib/upgrade/platform.sh:
```diff
@@ -24,6 +20,11 @@ platform_do_upgrade() {
+    buffalo_upgrade_prepare_ubi
+    CI_ROOTPART="ubi_rootfs"
+    nand_do_upgrade "$1"
+    ;;
```
Broken indent.

lmore377 commented:
Would it be alright if I made a new pull request? My branch is just sorta out of sync and I want to start fresh but I don't know how to just completely reset it and now every time I try to amend the commit it brings back the 4.19 files.

Ansuel commented:
@lmore377 the best you can do is just start fresh and then force push the branch to this...

lmore377 commented:
Yeah that's what I'm trying to do now.

adschm commented:
Please don't open a new Pull Request, as all the discussion so far is here.

Also note that after you have closed this PR now, you first have to go back to the state when you closed it, reopen the PR, and then force-push the changes to GitHub when the PR is open.

lmore377 commented:
Last time I left a comment it had the option to reopen so I'm just trying that rea quick

lmore377 commented:
Finally got it to reopen.

adschm commented:
Perfect. Now, keep it open, use "git rebase" and "git rebase -i ..." to rebase and squash your changes locally, and then force-push the result to GitHub ...

lmore377 committed:
ipq806x: add support for Askey RT4230W REV6

This adds support for the Askey RT4230W REV6
(Branded by Spectrum/Charter as RAC2V1K)

At this time, there's no way to reinstall the stock firmware so don't install
this on a router that's being rented.

Specifications:

    Qualcomm IPQ8065
    1 GB of RAM (DDR3)
    512 MB Flash (NAND)
    2x Wave 2 WiFi cards (QCA9984)
    5x 10/100/1000 Mbps Ethernet (Switch: QCA8337)
    1x LED (Controlled by a microcontroller that switches it between red and
blue with different patterns)
    1x USB 3.0 Type-A
    12V DC Power Input
    UART header on PCB - pinout from top to bottom is RX, TX, GND, 5V
    Port settings are 115200n8

More information: https://forum.openwrt.org/t/askey-rac2v1k-support/15830
https://deviwiki.com/wiki/Askey_RAC2V1K

To check what revision your router is, restore one of these config backups
through the stock firmware to get ssh access then run
"cat /proc/device-tree/model".
https://forum.openwrt.org/t/askey-rac2v1k-support/15830/17
The revision number on the board doesn't seem to be very consistent so that's
why this is needed. You can also run printenv in the uboot console and if
machid is set to 177d, that means your router is rev6.

Note: Don't install this if the router is being rented from an ISP. The defined
partition layout is different from the OEM one and even if you changed the
layout to match, backing up and restoring the OEM firmware breaks /overlay so
nothing will save and the router will likely enter a bootloop.

How to install:

Method 1: Install without opening the case using SSH and tftp

    You'll need:
    RAC2V1K-SSH.zip:
https://github.com/lmore377/openwrt-rt4230w/blob/master/RAC2V1K-SSH.zip
    initramfs and sysupgrade images

    Connect to one of the router's LAN ports

    Download the RAC2V1K-SSH.zip file and restore the config file that
corresponds to your router's firmware (If you're firmware is newer than what's
in the zip file, just restore the 1.1.16 file)

    After a reboot, you should be able to ssh into the router with username:
"4230w" and password: "linuxbox" or "admin". Run the following commannds
     fw_setenv ipaddr 10.42.0.10 #IP of router, can be anything as long as
it's in the same subnet as the server
     fw_setenv serverip 10.42.0.1# #IP of tftp server that's set up in next
steps
     fw_setenv bootdelay 8
     fw_setenv bootcmd "tftpboot initramfs.bin; bootm; bootipq"

    Don't reboot the router yet.

    Install and set up a tftp server on your computer

    Set a static ip on the ethernet interface of your computer (use this for
serverip in the above commands)

    Rename the initramfs image to initramfs.bin, and host it with the tftp
server

    Reboot the router. If you set up everything right, the router led should
switch over to a slow blue glow which means openwrt is booted. If for some
reason the file doesn't get loaded into ram properly, it should still boot to
the OEM firmware.
    After openwrt boots, ssh into it and run these commands:
    fw_setenv bootcmd "setenv mtdids nand0=nand0 && setenv mtdparts
 mtdparts=nand0:0x1A000000@0x2400000(firmware) && ubi part firmware && ubi
read 0x44000000 kernel 0x6e0000 && bootm"
    fw_setenv bootdelay 2

    After openwrt boots up, figure out a way to get the sysupgrade file onto it
(scp, custom build with usb kernel module included, wget, etc.) then flash it
with sysupgrade. After it finishes flashing, it should reboot, the light should
start flashing blue, then when the light starts "breathing" blue that means
openwrt is booted.

Method 2: Install with serial access (Do this if something fails and you can't
boot after using method 1)

    You'll need:
    initramfs and sysupgrade images
    Serial access:
https://openwrt.org/inbox/toh/askey/askey_rt4230w_rev6#opening_the_case

    Install and set up a tftp server

    Set a static ip on the ethernet interface of your computer

    Download the initramfs image, rename it to initramfs.bin, and host it with
the tftp server

    Connect the wan port of the router to your computer

    Interrupt U-Boot and run these commands:
    setenv serverip 10.42.0.1 (You can use whatever ip you set for the computer)
    setenv ipaddr 10.42.0.10 (Can be any ip as long as it's in the same subnet)
    setenv bootcmd "setenv mtdids nand0=nand0 &&
set mtdparts mtdparts=nand0:0x1A000000@0x2400000(firmware) && ubi part firmware
&& ubi read 0x44000000 kernel 0x6e0000 && bootm"

    saveenv
    tftpboot initramfs.bin
    bootm

    After openwrt boots up, figure out a way to get the sysupgrade file onto it
(scp, custom build with usb kernel module included, wget, etc.) then flash it
with sysupgrade. After it finishes flashing, it should reboot, the light should
start flashing blue, then when the light starts "breathing" blue that means
openwrt is booted.

Signed-off-by: Lauro Moreno <lmore377@gmail.com>

lmore377 commented:
I think everything's good now

lmore377 commented:
Can I get an update on the merge-ability of this? Also I still have no idea how to handle the led names.

lmore377 commented:
I just noticed that the status changed to those two red warnings. Is there anything I need to do on my end or is it fine?

adschm commented:
> I just noticed that the status changed to those two red warnings. Is there anything I need to do on my end or is it fine?

No, that's irrelevant, we won't merge via GitHub anyway. No idea why someone changed that, it hurts my eyes ;-)

lmore377 commented:
Any updates? @adschm @Ansuel

lmore377 commented:
@adschm @Ansuel I changed the mtd names to match the oem ones and I also moved setting the lan port mac addresses to 02_network. I've also been keeping up with making sure it should merge fine, even with the 2 new devices.

lmore377 commented:
@adschm @Ansuel Sorry that I keep mentioning you guys but it's been almost 4 months since this was last looked at. Can I get another review over everything?

lmore377 commented:
@adschm @Ansuel Maybe if I mention you guys during the day?

Ansuel commented:
We merged the asrock device and we skipped this wth ?

lmore377 commented:
I already accounted for that device in my repo

lmore377 commented:
@Ansuel Is there anything else that needs to be done or do I just keep waiting for a review?

lmore377 commented:
@Ansuel @adschm

blocktrron commented on target/linux/ipq806x/base-files/etc/board.d/02_network:
```diff
@@ -12,6 +12,12 @@ board_config_update
+askey,rt4230w-rev6)
+	ucidef_add_switch "switch0" \
+		"2:lan" "3:lan" "4:lan" "5:lan" "6@eth1" "1:wan" "0@eth0"
+	ucidef_set_interface_macaddr "lan" "$(mtd_get_mac_binary "0:ART" 0x6)"
+	ucidef_set_interface_macaddr "wan" "$(mtd_get_mac_binary "0:ART" 0x0)"
```
MAC addresses in device-tree are preferred, as it keeps the configuration portable across devices. Can you move them over there?

lmore377 replied:
Moved it

blocktrron commented on target/linux/ipq806x/base-files/lib/upgrade/platform.sh:
```diff
@@ -19,13 +29,9 @@ platform_do_upgrade() {
-	netgear,r7800 |\
-	qcom,ipq8064-ap148 |\
-	qcom,ipq8064-ap161)
+	buffalo,wxr-2533dhp)
+		buffalo_upgrade_prepare_ubi
+		CI_ROOTPART="ubi_rootfs"
 		nand_do_upgrade "$1"
```
A wild buffalo appears out of nowhere!

lmore377 replied:
Note to self: never trust that git rebase will do everything correctly

blocktrron commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,303 @@
+			gpios = <&qcom_pinmux 54 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+			debounce-interval = <60>;
+			wakeup-source;
```
AFAICT, this is not supported in the downstream driver. Given we won't be using the upstream driver, drop this please

lmore377 replied:
So I just completely delete these? What about the reset button? I didn't add any extra kmods or anything and they just work already so something's making it work.

mans0n replied:
You can just drop the `wakeup-source` line. The *downstream driver* refers to `kmod-gpio-button-hotplug`, which is added to the whole target by default. You can also drop the `debounce-interval` line unless you have a good reason to keep it. That's what I meant in https://github.com/openwrt/openwrt/pull/3115#discussion_r491644968.

lmore377 replied:
I removed the two lines @mans0n mentioned

blocktrron commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,303 @@
+			gpios = <&qcom_pinmux 68 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_WPS_BUTTON>;
+			debounce-interval = <60>;
+			wakeup-source;
```
same

blocktrron commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,303 @@
+		pinctrl-0 = <&led_pins>;
+		pinctrl-names = "default";
+
+		ledctrl1: ledctrl1 {
+			label = "rt4230w:ledctrl:1";
+			gpios = <&qcom_pinmux 22 GPIO_ACTIVE_HIGH>;
+		};
+		
+		ledctrl2: ledctrl2 {
+			label = "rt4230w:ledctrl:2";
+			gpios = <&qcom_pinmux 23 GPIO_ACTIVE_HIGH>;
+		};
+		
+		ledctrl3: ledctrl3 {
+			label = "rt4230w:ledctrl:3";
+			gpios = <&qcom_pinmux 24 GPIO_ACTIVE_HIGH>;
+		};
```
Do they all have the same color? Name should be set to "<color>:<indication>". Indication being for example "wps", "wifi", "eth"...

lmore377 replied:
I just left the names at ledctrl:1 thru 3. You can look at [this comment](https://github.com/openwrt/openwrt/pull/3115#discussion_r441098331) to see how the led on this device works.

blocktrron commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -92,6 +92,21 @@ define Device/ZyXELImage
+	PAGESIZE := 2048
+	DEVICE_PACKAGES := ath10k-firmware-qca9984-ct
+	KERNEL_IN_UBI := 1
+	IMAGES := nand-sysupgrade.bin
```
Not `nand-sysupgrade.bin` --> `sysupgrade.bin`

mans0n replied:
`Device/Default` already defines `sysupgrade.bin` so you can just drop the two lines. Or you can use `$(call Device/UbiFit)` as I mentioned in https://github.com/openwrt/openwrt/pull/3115#discussion_r447598459.

lmore377 replied:
Changing it from LegacyImage to UbiFit breaks compilation for some reason so I left it alone. I was able to remove the 2 lines without issues.

adschm commented:
Your commit message lacks a Signed-off-by. Apart from that, the text is really hard to read. Maybe consider improving your indenting scheme.

adschm commented on target/linux/ipq806x/base-files/etc/board.d/02_network:
```diff
@@ -84,3 +85,4 @@ esac
 board_config_flush
 
 exit 0
+
```
drop this empty line.

adschm commented on target/linux/ipq806x/base-files/lib/upgrade/platform.sh:
```diff
@@ -62,3 +63,4 @@ platform_do_upgrade() {
 		;;
 	esac
 }
+
```
drop this empty line.

adschm commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,303 @@
+		pinctrl-names = "default";
+
+		ledctrl1: ledctrl1 {
+			label = "ledctrl:1";
```
Remove the colon, "ledctrl" is not a color.

lmore377 replied:
When I change it to just `ledctrl1 {` I get this error for all of the leds
`arch/arm/boot/dts/qcom-ipq8065.dtsi:7.10-9.4: ERROR (path_references): /aliases: Reference to non-existent node or label "ledctrl3"`

adschm replied:
You are supposed to change the label property, not the DT labels in the line before.

adschm commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,303 @@
+			};
+			partition@2400000 {
+				label = "ubi";
+				reg = <0x2400000 0x1A000000>;
```
Please stick to lower case for the "A" in size here.

adschm commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts:
```diff
@@ -0,0 +1,303 @@
+			0x7c 0x4e 
+			0x94 0x4e 
+			0x50 0xcf02cf02 
+			0x54 0xc832c832
```
It would be nice to add some comments what these registers are (like in most other files).

adschm commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -92,6 +92,19 @@ define Device/ZyXELImage
+        BLOCKSIZE := 128k
+        PAGESIZE := 2048
+        DEVICE_PACKAGES := ath10k-firmware-qca9984-ct
+        KERNEL_IN_UBI := 1
```
Indent should be tabs, not spaces here.

adschm commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -411,3 +424,4 @@ endef
 TARGET_DEVICES += zyxel_nbg6817
 
 $(eval $(call BuildImage))
+
```
Remove empty line here.

adschm commented on target/linux/ipq806x/patches-5.4/0069-arm-boot-add-dts-files.patch:
```diff
@@ -29,9 +29,11 @@ Signed-off-by: John Crispin <john@phrozen.org>
  	qcom-msm8660-surf.dtb \
  	qcom-msm8960-cdp.dtb \
  	qcom-msm8974-fairphone-fp2.dtb \
+
```
Remove empty line here.

lmore377 replied:
Sorry about all these weird things. gedit just likes to mess with things for some reason.

lmore377 commented:
I've been busy with school so I can't work on this much. I'll update when things slow down a bit.

lmore377 commented:
@adschm Is is there anything else I need to do to get this accepted?

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

ynezz commented:
> Is is there anything else I need to do to get this accepted?

Yes, add your Signed-off-by to the commit message.

lmore377 commented:
> Yes, add your Signed-off-by to the commit message.

Consider it done

lmore377 commented:
@ynezz I added the Signed-off-by to the commit message

adschm commented:
Since @ynezz already had it in his staging tree and the remaining issue has been fixed, I'll merge it now with a few whitespace fixes and added support for 5.10 ...

lmore377 commented:
Awesome. Is there anything else I need to do from this point on?

adschm commented:
Well, if you find there is anything broken or incorrect, you are encouraged to still provide a fix now :-)

lmore377 commented:
Alright sounds good. Is there a process to get it into stable or does it just kind of happen at some point?

adschm commented:
It will get into next stable when it is branched automatically.
